File: src/main/java/com/github/ofrostdev/api/utils/TitleAPI.java
package com.github.ofrostdev.api.utils;

import org.bukkit.entity.Player;
import org.bukkit.Bukkit;

import java.lang.reflect.Constructor;
import java.lang.reflect.Method;

public class TitleAPI {

    public static void send(Player player, String title, String subtitle, int fadeIn, int stay, int fadeOut) {
        try {
            String version = Bukkit.getServer().getClass().getPackage().getName().split("\\.")[3];
            Class<?> chatComponentText = Class.forName("net.minecraft.server." + version + ".ChatComponentText");
            Class<?> iChatBaseComponent = Class.forName("net.minecraft.server." + version + ".IChatBaseComponent");

            Object titleComponent = chatComponentText.getConstructor(String.class).newInstance(title);
            Object subtitleComponent = chatComponentText.getConstructor(String.class).newInstance(subtitle);

            Class<?> packetPlayOutTitle = Class.forName("net.minecraft.server." + version + ".PacketPlayOutTitle");
            Class<?> enumTitleAction = Class.forName("net.minecraft.server." + version + ".PacketPlayOutTitle$EnumTitleAction");

            Object titlePacket = packetPlayOutTitle.getConstructor(enumTitleAction, iChatBaseComponent)
                    .newInstance(Enum.valueOf((Class<Enum>) enumTitleAction, "TITLE"), titleComponent);

            Object subtitlePacket = packetPlayOutTitle.getConstructor(enumTitleAction, iChatBaseComponent)
                    .newInstance(Enum.valueOf((Class<Enum>) enumTitleAction, "SUBTITLE"), subtitleComponent);

            Constructor<?> timingConstructor = packetPlayOutTitle.getConstructor(int.class, int.class, int.class);
            Object timingPacket = timingConstructor.newInstance(fadeIn, stay, fadeOut);

            sendPacket(player, timingPacket);
            sendPacket(player, titlePacket);
            sendPacket(player, subtitlePacket);

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void send(Player player, String title, int fadeIn, int stay, int fadeOut) {
        send(player, title, "", fadeIn, stay, fadeOut);
    }

    private static void sendPacket(Player player, Object packet) {
        try {
            String version = Bukkit.getServer().getClass().getPackage().getName().split("\\.")[3];
            Class<?> craftPlayerClass = Class.forName("org.bukkit.craftbukkit." + version + ".entity.CraftPlayer");
            Object craftPlayer = craftPlayerClass.cast(player);
            Object handle = craftPlayerClass.getMethod("getHandle").invoke(craftPlayer);
            Object playerConnection = handle.getClass().getField("playerConnection").get(handle);
            Method sendPacket = playerConnection.getClass().getMethod("sendPacket", Class.forName("net.minecraft.server." + version + ".Packet"));
            sendPacket.invoke(playerConnection, packet);
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/github/ofrostdev/api/utils/packets/common/TitleAPI.kt
package com.github.ofrostdev.api.utils.packets.common

import com.github.ofrostdev.api.utils.packets.PacketUtils
import net.minecraft.server.v1_8_R3.IChatBaseComponent
import net.minecraft.server.v1_8_R3.IChatBaseComponent.ChatSerializer
import net.minecraft.server.v1_8_R3.PacketPlayOutTitle
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Sound
import org.bukkit.entity.Player
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitRunnable
import java.util.*
import java.util.concurrent.ConcurrentHashMap

@DslMarker
annotation class TitleDSL

@TitleDSL
class TitleAPI private constructor(
    private val plugin: Plugin,
    private var title: String,
    private var subtitle: String? = null
) {
    private var sound: Sound? = null
    private val placeholders: MutableMap<String, String> = mutableMapOf()

    fun replace(target: String, replacement: String) = apply { placeholders[target] = replacement }
    fun replaceAll(replacements: Map<String, String>) = apply { placeholders.putAll(replacements) }

    fun colorize() = apply {
        title = ChatColor.translateAlternateColorCodes('&', title)
        subtitle = subtitle?.let { ChatColor.translateAlternateColorCodes('&', it) }
    }

    fun sound(sound: Sound) = apply { this.sound = sound }

    private fun applyPlaceholders(text: String?): String? {
        var result = text
        placeholders.forEach { (key, value) -> result = result?.replace(key, value) }
        return result
    }

    fun send(player: Player, fadeIn: Int = 10, stay: Int = 70, fadeOut: Int = 20) {
        if (!player.isOnline) return
        val t = applyPlaceholders(title) ?: ""
        val s = applyPlaceholders(subtitle) ?: ""

        val titleComponent: IChatBaseComponent = ChatSerializer.a("{\"text\":\"$t\"}")
        val subtitleComponent: IChatBaseComponent = ChatSerializer.a("{\"text\":\"$s\"}")

        val packetTitle = PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction.TITLE, titleComponent, fadeIn, stay, fadeOut)
        val packetSubtitle = PacketPlayOutTitle(PacketPlayOutTitle.EnumTitleAction.SUBTITLE, subtitleComponent, fadeIn, stay, fadeOut)

        PacketUtils.sendPacket(player, packetTitle)
        PacketUtils.sendPacket(player, packetSubtitle)
        sound?.let { player.playSound(player.location, it, 1.0f, 1.0f) }
    }

    fun send(players: Collection<Player>, fadeIn: Int = 10, stay: Int = 70, fadeOut: Int = 20) {
        players.forEach { send(it, fadeIn, stay, fadeOut) }
    }

    fun broadcast(fadeIn: Int = 10, stay: Int = 70, fadeOut: Int = 20) {
        send(Bukkit.getOnlinePlayers(), fadeIn, stay, fadeOut)
    }

    companion object {
        private lateinit var pluginInstance: Plugin
        private val activeTasks: MutableMap<UUID, BukkitRunnable> = ConcurrentHashMap()

        fun enable(plugin: Plugin) {
            pluginInstance = plugin
        }

        fun of(title: String, subtitle: String? = null): TitleAPI {
            check(::pluginInstance.isInitialized) { "[FrostAPI] TitleAPI instance is null!" }
            return TitleAPI(pluginInstance, title, subtitle)
        }

        fun of(plugin: Plugin, title: String, subtitle: String? = null): TitleAPI {
            return TitleAPI(plugin, title, subtitle)
        }
    }
}

fun TitleAPI.dsl(block: TitleAPI.() -> Unit): TitleAPI {
    this.block()
    return this
}

fun title(title: String, subtitle: String? = null, block: TitleAPI.() -> Unit): TitleAPI {
    return TitleAPI.of(title, subtitle).dsl(block)
}

fun title(plugin: Plugin, title: String, subtitle: String? = null, block: TitleAPI.() -> Unit): TitleAPI {
    return TitleAPI.of(plugin, title, subtitle).dsl(block)
}
